lib/actions/post.actions.ts

```diff
@@ -68,13 +98,40 @@ export async function getAllPosts({
         userEmail,
         ...(searchTerm &&
           searchTerm !== "all" && { createType: searchTerm as CreateType }),
+        ...(term && {
+          OR: [
+            {
+              title: {
+                contains: term,
+                mode: "insensitive",
+              },
+            },
+            {
+              content: {
+                contains: term,
+                mode: "insensitive",
+              },
+            },
+            {
+              description: {
+                contains: term,
+                mode: "insensitive",
+              },
+            },
+            {
+              tags: {
+                has: term,
+              },
+            },
+          ],
+        }),
```

Comment:
there's a bunch of duplicated code here...

Reply:
Yes there is. Remember these are objects. So you can make an object for these two functions to share.

What have you tried? 

The reality is that you can either make two different queries here (one to count the posts with the filter, and one to select just some of the results with that filter) or you can make 1 query but return EVERY result and just count them.

I would recommend the method you currently have as it scales. Returning all the data for every post for every request is too much, so an aggregate like you have here makes sense.

There are ways to simplify the code though if you want to give it a try!

Firstly you should identify what's being reused between both queries.

In this case, it's the 'where' statement.

```js
 where: {
        userEmail,
        ...(searchTerm &&
          searchTerm !== "all" && { createType: searchTerm as CreateType }),
        ...(term && {
          OR: [
            {
              title: {
                contains: term,
                mode: "insensitive",
              },
            },
            {
              content: {
                contains: term,
                mode: "insensitive",
              },
            },
            {
              description: {
                contains: term,
                mode: "insensitive",
              },
            },
            {
              tags: {
                has: term,
              },
            },
          ],
        }),
      },
  ```


So you can make a reusable 'options' object that contains these values (remember these are just objects. You can create them and generate them as you see fit. That's just javascript)

So 

```js
    // declare the reused options object
    const options = {
      where: {
        userEmail,
        ...(searchTerm &&
          searchTerm !== "all" && { createType: searchTerm as CreateType }),
        ...(term && {
          OR: [
            { title: { contains: term, mode: "insensitive" } },
            { content: { contains: term, mode: "insensitive" } },
            { description: { contains: term, mode: "insensitive" } },
            { tags: { has: term } },
          ],
        }),
      },
    };
```

Now your 'findMany' can become:
```js
      prisma.post.findMany({
        ...(options as Prisma.PostFindManyArgs),
        orderBy: [{ createdAt: "desc" }],
        skip: (Number(page) - 1) * postsToTake,
        take: postsToTake + 1,
      }),
```

and your aggregate can become:
```js
      prisma.post.aggregate({
        _count: true,
        ...(options as Prisma.PostAggregateArgs),
      }) as any,
```

That reduces some code repetition. 

```js
export async function getAllPosts({
  page,
  searchTerm,
  postsToTake = 4,
  term,
}: {
  page: string;
  searchTerm?: string;
  postsToTake?: number;
  term?: string;
}) {
  let hasNextPage = false;
  try {
    const userEmail = await getUserSession();
    // declare the reused options object
    const options = {
      where: {
        userEmail,
        ...(searchTerm &&
          searchTerm !== "all" && { createType: searchTerm as CreateType }),
        ...(term && {
          OR: [
            { title: { contains: term, mode: "insensitive" } },
            { content: { contains: term, mode: "insensitive" } },
            { description: { contains: term, mode: "insensitive" } },
            { tags: { has: term } },
          ],
        }),
      },
    };

    // use that object to make both queries
    const somePosts = await prisma.post.findMany({
      ...(options as Prisma.PostFindManyArgs),
      orderBy: [{ createdAt: "desc" }],
      skip: (Number(page) - 1) * postsToTake,
      take: postsToTake + 1,
    });

    const { _count: numberOfResults } = (await prisma.post.aggregate({
      _count: true,
      ...(options as Prisma.PostAggregateArgs),
    })) as any;

    if (somePosts.length > postsToTake) {
      somePosts.pop();
      hasNextPage = true;
    }
    return {
      somePosts,
      hasNextPage,
      numberOfPages: Math.ceil(numberOfResults / postsToTake) || 1,
    };
  } catch (error) {
    console.error("Error returning posts:", error);
    return { error: "An unexpected error occurred while returning posts." };
  }
}
```